ops/model.py

```diff
@@ -691,6 +691,32 @@ def set_ports(self, *ports: Union[int, 'Port']) -> None:
         for protocol, port in desired - existing:
             self._backend.open_port(protocol, port)
 
+    def reboot(self, now: bool = False) -> None:
+        """Reboot the host machine, after stopping all containers hosted on the machine.
+
+        Normally, the reboot will only take place after the current hook successfully
+        completes. Use the ``now`` argument when multiple reboots are required, to
+        reboot immediately without waiting for the hook to complete, and to restart the
+        hook after rebooting.
+
+        This will silently fail for Kubernetes charms.
```

Comment:
Hmm, yeah, that's a pain. Can we make it loudly fail? It sounds like the Juju hook tool doesn't fail with an exit code right now. What's the error message it logs? If it's not something *too* generic we could look for that error message on stdout (or is it stderr?) and raise a nice error. It's normally not great to parse error messages like this in case the Juju side changes, but all we're doing is improving the error handling, so I'd be in favour of that. (We do this in a couple of other places, for example `_ModelBackend._is_relation_not_found`.)

Reply:
The error is in the logger, ie. `debug-log` shows this:

```
unit-uptime-0: 12:59:56 ERROR juju.worker.uniter.operation hook "config-changed" (via hook dispatching script: dispatch) failed: unit "uptime/0" is not assigned to a machine
```

It indeed does not fail with an exit code (or `ModelError` would conveniently be raised, as with running in an action hook). When passing `return_output=True` to `_run()` I get an empty string back, so I believe that means that it's not putting anything into stdout/stderr. I was assuming we would not want to introspect the Juju logs.

Maybe Juju should be exiting with an error in this case? I could ask the Juju team about that and open a PR for it if so.

Reply:
Ah, got it. Yeah, we definitely don't want to introspect Juju logs.

However, it looks like `_run()` doesn't return stderr (unless there's an exception / nonzero exit code). Perhaps you can try running it directly using `subprocess.run` and print both stdout and stderr to see if we get any stderr.

I definitely think Juju should be returning an error in this case. We'll decide whether it's worth opening an issue there after we determine the above.

Reply:
`stderr` is empty as well :disappointed:.

Reply:
Oh, I bet I know why -- it's because it's not doing it *now* and so the hook goes through fine, and it's only doing the check at the end of the hook when the reboot is "committed". We ran into this with something else too. Ideally Juju would do the "is this supported on this environment" check early, but oh well.

I bet that `reboot(now=True)` on K8s will give the error right away.

In any case, for now I'd say let's change the comment from "This will silently fail for Kubernetes charms" ("silently fail" is a bit odd) to "This is not supported on Kubernetes charms" and just let them deal.

Reply:
> I bet that `reboot(now=True)` on K8s will give the error right away.

It's difficult to tell, because the Charm is killed before we get back from subprocess.call when `--now` is used.

> In any case, for now I'd say let's change the comment from "This will silently fail for Kubernetes charms" ("silently fail" is a bit odd) to "This is not supported on Kubernetes charms" and just let them deal.

Sure.